### DISCUSSION TITLE
What is the best practice to get scoped service from singletone service

### DISCUSSION THREAD

**Discussion by HengzheLi** [OrchardCMS/OrchardCore#10686] [Q&A] [answered]
Some times when I need a scoped service from a singletone service, I have to manually get the tenant level service provider first using below code
```
var shell = rootProvider.GetRequiredService<IShellHost>();
var sp = shell.GetScopeAsync(root.GetRequiredService<ShellSettings>()).ConfigureAwait(false)
                            .GetAwaiter().GetResult().ServiceProvider;
```
Then use the tenant level service provider to get scoped service.

It looks verbose. Has a way to inject the tenant level service provider into constructor then it can be used directly?

**Comment by jtkech** [comment, marked as answer]
If you are sure that your method is called in a scope context

    using Microsoft.Extensions.DependencyInjection;
    using OrchardCore.Environment.Shell.Scope;

            var service = ShellScope.Services.GetRequiredService<IService>();

In a more aspnetcore way you can inject `IHttpContextAccessor` to retrieve the current `RequestServices`. A request context is not the same as a scope context (here the scope of a given tenant / shell) but we take care in a given request that `RequestServices` is set to the scope service provider of the tenant / shell on which the request was done.

Otherwise if the method is not called in a scope context you should create a scope by using `GetScopeAsync()`, notice that in your singleton (if tenant singleton not app level one) you can inject `ShellSettings` that is a tenant singleton. Also make your method async to not have to use `GetAwaiter()`, under aspnetcore no need for `ConfigureAwait()`. Then in this context you should use the shell scope by using our helper `scope.UsingAsync()`, the result being something like this

    var scope await _shellHost.GetScopeAsync(_shellSettings);
    await scope.UsingAsync(async scope =>
    {
        var service = scope.Services.GetRequiredService<IService>();
        ...
    });
